server/etcdserver/api/v3rpc/qmon.go

```diff
@@ -0,0 +1,357 @@
+	}
+	if ctrl.budgetExhausted {
+		//decline
+		return true, respSize, qcount
+	}
+
+	//admit
+	return false, respSize, qcount
+}
+
+func (ctrl *BandwidthMonitor) newQueryFromReqResp(req interface{}, resp interface{}) Query {
+	var reqSize, respSize int
+	var reqContent string
+	qtype := QueryTypeUnknown
+	switch _resp := resp.(type) {
+	case *pb.RangeResponse:
+		_req, ok := req.(*pb.RangeRequest)
+		if ok && !_req.CountOnly {
+			reqSize = _req.Size()
+			reqContent = _req.String()
+			qtype = QueryTypeRange
+		}
+		if _resp != nil && _resp.Count != 0 {
+			respSize = _resp.Size()
+		}
+	default:
+		reqSize = 0
+		respSize = 0
+	}
+
+	var q Query
+	q.qid = reqContent
+	q.qsize = uint64(reqSize + respSize)
+	q.qtype = qtype
+	return q
+}
+
+func (ctrl *BandwidthMonitor) newQueryFromReq(req interface{}) Query {
+	var q Query
+	switch req.(type) {
+	case *pb.RangeRequest:
+		_req, ok := req.(*pb.RangeRequest)
+		if ok {
+			q = ctrl.newQRange(_req)
+		}
+	default:
+	}
+	return q
+}
+
+func (ctrl *BandwidthMonitor) newQRange(r *pb.RangeRequest) Query {
+	var q Query
+	q.qid = r.String()
+	q.qsize = 0
+	q.qtype = QueryTypeRange
+	return q
+}
+
+func (ctrl *BandwidthMonitor) UpdateUsage(req interface{}, resp interface{}, err error) {
+
+	//do not update estimate with failure response
+	if err != nil {
+		return
+	}
+
+	q := ctrl.newQueryFromReqResp(req, resp)
+	ctrl.mu.Lock()
+	defer ctrl.mu.Unlock()
+	if q.qtype == QueryTypeUnknown {
+		return
+	}
+
+	current := ctrl.estRespSize.EstimateString(q.qid)
+	if current == 0 {
+		ctrl.estRespSize.UpdateString(q.qid, q.qsize)
+		current = q.qsize
+	}
+	if current != q.qsize && q.qsize > SmallReqThreshold {
+		ctrl.server.Cfg.Logger.Warn("qmon qsize changed. update estimates.", zap.String("qid", q.qid), zap.Uint64("qsize", q.qsize), zap.Uint64("current", current))
+		ctrl.updateEstimate = true
+	}
+	if ctrl.auditOn {
+		ctrl.server.Cfg.Logger.Warn("qmon audit.", zap.String("qid", q.qid), zap.Uint64("qsize", q.qsize))
+	}
+}
+
+func (ctrl *BandwidthMonitor) AdmitReq(req interface{}) bool {
+	q := ctrl.newQueryFromReq(req)
+	declined, qsize, qcount := ctrl.isDeclined(q)
+
+	//dont rely on default resp size too much
+	if (ctrl.alwaysOnForLargeReq || declined) && qcount > 5 && qsize == ctrl.defaultRespSize {
+		ctrl.server.Cfg.Logger.Warn("qmon reject. Response size unknown.", zap.String("qid", q.qid), zap.Uint64("qsize", qsize), zap.Uint64("qcount", qcount))
+		return false
+	}
+
+	if (qsize > SmallReqThreshold && declined) || (qsize > LargeReqThreshold && ctrl.alwaysOnForLargeReq) {
+		//TODO add metric
+		err := ctrl.throttle.WaitN(context.TODO(), int(qsize))
+		if err != nil {
+			return false
+		}
+	}
+	return true
+}
+
+func (ctrl *BandwidthMonitor) isDeclined(q Query) (bool, uint64, uint64) {
+	ctrl.mu.Lock()
+	defer ctrl.mu.Unlock()
+	return ctrl.isDeclinedUnsafe(q)
+}
+
+func (ctrl *BandwidthMonitor) Stop() {
+}
+
+//TODO windows
+func getCurrentRssBytes(logger *zap.Logger) uint64 {
```

Comment:
we may get the process_rss prometheus metric from the registry directly and it should also work in other platforms.

Reply:
ack. I will look into this one incrementally.